packages/datadog-plugin-aws-sdk/test/kinesis.spec.js

```diff
@@ -35,117 +43,210 @@ describe('Kinesis', () => {
-        expect(span.meta).to.have.property('streamname', 'MyStream')
-      }).then(done, done)
 
-      helpers.putTestRecord(kinesis, helpers.dataBuffer, e => e && done(e))
+        it('skip injects trace context to Kinesis putRecord when disabled', done => {
+          helpers.putTestRecord(kinesis, streamName, helpers.dataBuffer, (err, data) => {
+            if (err) return done(err)
+
+            helpers.getTestData(kinesis, streamName, data, (err, data) => {
+              if (err) return done(err)
+
+              expect(data).not.to.have.property('_datadog')
+
+              done()
+            })
+          })
+        })
+      })
     })
 
-    describe('Disabled', () => {
+    describe('DSM Context Propagation', () => {
       before(() => {
-        process.env.DD_TRACE_AWS_SDK_KINESIS_ENABLED = 'false'
+        return agent.load('aws-sdk', { kinesis: { dsmEnabled: true } }, { dsmEnabled: true })
       })
 
-      after(() => {
-        delete process.env.DD_TRACE_AWS_SDK_KINESIS_ENABLED
+      before(done => {
+        tracer = require('../../dd-trace')
+        tracer.use('aws-sdk', { kinesis: { dsmEnabled: true } }, { dsmEnabled: true })
+
+        createResources(streamNameDSM, done)
+      })
+
+      after(done => {
+        kinesis.deleteStream({
+          StreamName: streamNameDSM
+        }, (err, res) => {
+          if (err) return done(err)
+
+          helpers.waitForDeletedStream(kinesis, streamNameDSM, done)
+        })
       })
 
-      it('skip injects trace context to Kinesis putRecord when disabled', done => {
-        helpers.putTestRecord(kinesis, helpers.dataBuffer, (err, data) => {
+      it('injects DSM pathway hash during Kinesis putRecord to the span', done => {
+        helpers.putTestRecord(kinesis, streamNameDSM, helpers.dataBuffer, (err, data) => {
           if (err) return done(err)
 
-          helpers.getTestData(kinesis, data, (err, data) => {
+          let putRecordSpanMeta = {}
+          agent.use(traces => {
+            const span = traces[0][0]
+
+            if (span.resource.startsWith('putRecord')) {
+              putRecordSpanMeta = span.meta
+            }
+
+            expect(putRecordSpanMeta).to.include({
+              'pathway.hash': expectedProducerHash.readBigUInt64BE(0).toString()
+            })
+          }).then(done, done)
+        })
+      })
+
+      describe('emits a new DSM Stats to the agent when DSM is enabled', () => {
+        before(done => {
+          helpers.putTestRecord(kinesis, streamNameDSM, helpers.dataBuffer, (err, data) => {
             if (err) return done(err)
 
-            expect(data).not.to.have.property('_datadog')
+            helpers.getTestData(kinesis, streamNameDSM, data, (err, data) => {
+              if (err) return done(err)
 
-            done()
+              tracer._tracer._dataStreamsProcessor.onInterval()
+
+              const intervalId = setInterval(() => {
+                if (agent.getDsmStats().length >= 1) {
```

Comment:
Let's not rely on a timer for this and instead have some sort of callback mechanism, similar to `agent.use`.

Reply:
I restructured to use a callback with `agent.use`

Reply:
The new approach looks fine, but let's have an explicit method instead of using `.use` directly. That method is only for traces, and should probably be renamed something like `expectTrace` at this point.

Reply:
Updated this code to use a separate method called `agent.expectStats`, which will only run against stats payloads.